### PR TITLE
mixer: add size function to EL

### PR DESCRIPTION
**Pull request by kyessenov**
Fixes https://github.com/istio/istio/issues/10011

Adds a polymorphic intrinsic function `size` (common with CEL). Implement it for strings.
It turns out that extending this function to stringmap is rather difficult:
- stringmap type is erased in bytecode, so it requires an extern fallback for reflection
- reference tracking for map size requires special handling in the proxy

Due to the above, I left out stringmap size implementation for nebulous future.

/assign @mandarjog 
/assign @ozevren 

Signed-off-by: Kuat Yessenov <kuat@google.com>

### CODE REVIEW

File: mixer/pkg/il/interpreter/interpreterRun.go [hunk 1]
```diff
@@ -868,6 +868,21 @@ func (in *Interpreter) run(fn *il.Function, bag attribute.Bag, step bool) (Resul
...
+		case il.SizeS:
+			if sp < 1 {
+				goto STACK_UNDERFLOW
+			}
+			t1 = opstack[sp-1]
+			sp = sp - 1
+			if t1 >= hp {
+				goto INVALID_HEAP_ACCESS
+			}
+			tStr = heap[t1].(string)
```
Comment (mandarjog) [inline mixer/pkg/il/interpreter/interpreterRun.go]: Panics if wrong type?

File: mixer/pkg/il/interpreter/interpreterRun.go [hunk 2]
```diff
@@ -868,6 +868,21 @@ func (in *Interpreter) run(fn *il.Function, bag attribute.Bag, step bool) (Resul
 			opstack[sp+1] = uint32(tu64 & 0xFFFFFFFF)
 			sp = sp + 2
 
+		case il.SizeS:
```
[the remark this answers is not in the document]
Comment (kyessenov) [reply mixer/pkg/il/interpreter/interpreterRun.go]: ah, thanks for pointing out

File: mixer/pkg/il/interpreter/interpreterRun.go (hunk 1 above)
Comment (kyessenov) [reply mixer/pkg/il/interpreter/interpreterRun.go]: it's type checked during translation.

File: mixer/pkg/lang/ast/func.go [hunk 3]
```diff
@@ -78,6 +78,11 @@ func intrinsics() []FunctionMetadata {
 			ReturnType:    config.VALUE_TYPE_UNSPECIFIED,
 			ArgumentTypes: []config.ValueType{config.VALUE_TYPE_UNSPECIFIED, config.VALUE_TYPE_UNSPECIFIED},
 		},
+		{
+			Name:          "size",
+			ReturnType:    config.INT64,
+			ArgumentTypes: []config.ValueType{config.VALUE_TYPE_UNSPECIFIED},
```
[the remark this answers is not in the document]
Comment (kyessenov) [reply mixer/pkg/lang/ast/func.go]: Sure.

File: mixer/pkg/il/interpreter/interpreterRun.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ozevren) [reply mixer/pkg/il/interpreter/interpreterRun.go]: Apart from this point, looks good to me.

File: mixer/pkg/il/interpreter/interpreterRun.got [hunk 4]
```diff
@@ -635,6 +635,13 @@ func (in *Interpreter) run(fn *il.Function, bag attribute.Bag, step bool) (Resul
 			tu64 = math.Float64bits(tf64)
 			STACK_PUSH2(uint32(tu64>>32), uint32(tu64&0xFFFFFFFF))
 
+		case il.SizeS:
```
Comment (ozevren) [inline mixer/pkg/il/interpreter/interpreterRun.got]: Actually, can you add stack underflow guard test for this as well?
It should be TestInterpreter_Eval_StackUnderflow

Comment (kyessenov) [reply mixer/pkg/il/interpreter/interpreterRun.got]: Will do in a followup.

**Pull request merged**